### ISSUE TITLE
PriorBox layer Backpropagation

### ISSUE THREAD

**Issue by leonardoaraujosantos** [weiliu89/caffe#387]
Hi @weiliu89 inspecting the code of the [Priorbox ](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/prior_box_layer.cpp#L120)layer I got confused on these topics:

1. There is no backpropagation code? The priorbox does not participate on the training?
2. There is no .cu version? It will not hurt performance?
3. The main idea of this layer is to generate all the possible bounding box position on each feature map?

**Comment by weiliu89** [comment]
1. PriorBox layer only generates prior box which is used to compute gradient in MultiBoxLayer.
2. It is not very computational expensive. Of course, you can optimize the speed if you like.
3. Yes.

**Comment by leonardoaraujosantos** [comment]
Hi @weiliu89 , thanks for the explanation, so here I added a diagram with the Priorbox inside the Detection part of SSD.
### Diagram
![ssd_detection](https://cloud.githubusercontent.com/assets/898383/21889773/7eac3544-d8c2-11e6-89a3-66b7b04faab0.png)

### My observations (Please correct if I'm wrong)
1. The priorbox layer during training is used  the matching between the Ground-Truth and it's default boxes. (Using Jaccard-Intersection over union)
2. It's not clear for me yet what the priorbox does on the prediction phase. (Question)

Looking on the priorbox layer forward cpu code it seems that is just creating the default boxes and clipping the default boxes that are out of the window.
![output_nmfojz](https://cloud.githubusercontent.com/assets/898383/21890123/1cc61d34-d8c4-11e6-9204-9fe1bcc9db07.gif)

Thanks again for the support

**Comment by weiliu89** [comment]
1. During training, prior box is used to match ground truth box, and is also used to encode ground truth box to compute the localization loss.

2. During inference, prior box is used in DetectionOutputLayer, and is used to decode the final detection outputs (apply location offset prediction on top of prior box).
